src/main/scala/com/microsoft/hyperspace/index/zordercovering/ZOrderField.scala

```diff
@@ -0,0 +1,569 @@
+
+      if (valueStr > maxVal || valueStr < minVal) {
+        throwOutOfRangeException(minVal, maxVal, valueStr)
+      }
+      val value = valueStr.substring(firstDiffIdx.min(valueStr.length))
+      val len = math.min(byteLen, value.length)
+      val padArr = Seq.fill(byteLen - len) { 0.toByte }
+      // For the first diff char, use (the char - minFirstDiff)
+      // to remove same upper bit sequence.
+      val firstByte = if (value.nonEmpty) {
+        (value.head - minFirstDiff).toByte
+      } else {
+        0.toByte
+      }
+      BitSet.valueOf((firstByte +: (value.slice(1, len).getBytes ++ padArr)).reverse)
+    }
+  }
+}
+
+/**
+ * For Timestamp type, handle it as Long type.
+ * See [[LongMinMaxZOrderField]].
+ *
+ * TODO: Support PercentileBasedZOrderField for Timestamp and DateType.
+ *   Spark cannot calculate quantiles for non-numeric type, need to cast to Long type
+ *   before collecting stats.
+ */
+case class TimestampMinMaxZOrderField(
+    override val name: String,
+    minVal: java.sql.Timestamp,
+    maxVal: java.sql.Timestamp)
+    extends ZOrderField {
+
+  val longField =
+    LongMinMaxZOrderField(
+      name,
+      util.DateTimeUtils.fromJavaTimestamp(minVal),
+      util.DateTimeUtils.fromJavaTimestamp(maxVal))
+
+  override val bitLen: Int = {
+    longField.bitLen
+  }
+
+  override def getBitSet(valueAny: Any): BitSet = {
+    if (checkEmptySetEligible(valueAny)) {
+      emptyBitSet
+    } else {
+      val value = valueAny.asInstanceOf[java.sql.Timestamp]
+      if (value.compareTo(minVal) < 0 || value.compareTo(maxVal) > 0) {
+        throwOutOfRangeException(minVal, maxVal, value)
+      }
+      val valueTime = util.DateTimeUtils.fromJavaTimestamp(value)
+      longField.getBitSet(valueTime)
+    }
+  }
+}
+
+/**
+ * For Date type, handle it as Int type after converting it as the number of days.
+ * See [[IntZOrderField]].
+ */
+case class DateMinMaxZOrderField(
+    override val name: String,
+    minVal: java.sql.Date,
+    maxVal: java.sql.Date)
+    extends ZOrderField {
+
+  val intField =
+    IntMinMaxZOrderField(
+      name,
+      util.DateTimeUtils.fromJavaDate(minVal),
+      util.DateTimeUtils.fromJavaDate(maxVal))
+
+  override val bitLen: Int = {
+    intField.bitLen
+  }
+
+  override def getBitSet(valueAny: Any): BitSet = {
+    if (checkEmptySetEligible(valueAny)) {
+      emptyBitSet
+    } else {
+      val value = valueAny.asInstanceOf[java.sql.Date]
+      if (value.compareTo(minVal) < 0 || value.compareTo(maxVal) > 0) {
+        throwOutOfRangeException(minVal, maxVal, value)
+      }
+      val valueDay = util.DateTimeUtils.fromJavaDate(value)
+      intField.getBitSet(valueDay)
+    }
+  }
+}
+
+object ZOrderField {
+  def build(
+      name: String,
+      dataType: DataType,
+      minVal: Any,
+      maxVal: Any,
+      quantiles: Seq[Any] = Nil,
+      quantileEnabled: Boolean = false): ZOrderField = {
+    dataType match {
+      case LongType =>
```

Comment:
does it really have to go through the list, or can the type be extracted from dataType parameter for brevity?

Reply:
I don't think the implementation is inefficient and it would be okay as this is one time cost for each z-order columns.
Could you elaborate more if it still needs to be improved?

Reply:
I think @MironAtHome is saying about code repetition. Having a long list of `if` or `case` conditions with similar code modulo some parameter is a typical antipattern. If it were C++, this kind of repetition can be easily factored out using templates, but in Scala, I think it needs more boilerplates and glue code than necessary.

We could have done something like this:
```scala
val builder: {def apply(name: String, min: Any, max: Any): ZOrderField} = dataType match {
  case IntegerType => IntMinMaxZOrderField
  case LongType => LongMinMaxZOrderField
  ...
}
builder(name, min, max)
```

But the actual code is slightly more complex than the simple `apply` function above and we would need to define a factory trait and factories for each class anyway, which seems to be a little too much considering there will be very few, if any, new types to Spark in the future, let alone whether we will support them or not.

I do prefer a modular approach where I would group classes by their differentiating properties (so I would put IntMinMaxZOrderField, IntPercentileZOrderField, and factories for them in IntZOrderField.scala, and so on) and make the combining code like `ZOrderField.build` agnostic about different implementations, meaning that you don't have to change the function when you add/remove implementations. But such a modular approach makes more sense for a larger abstraction. If we need to, we can change to that later.